### ISSUE TITLE
torch._dynamo.optimize loses the setting of requires_grad

### ISSUE THREAD

**Issue by CaoE** [pytorch/pytorch#107861]
### 🐛 Describe the bug

The issue is exposed by https://github.com/pytorch/pytorch/actions/runs/5917745793/job/16046614052.
When torch._dynamo.optimize is enabled for the code shown below,  the setting of `requires_grad` is lost.

```
import torch

def traverse_obj(obj, func):
    if isinstance(obj, (tuple, list)):
        a = type(obj)(traverse_obj(o, func) for o in obj)
        print("return tuple tensor: ", a[0].requires_grad)
        return a
    elif isinstance(obj, dict):
        return {name: traverse_obj(o, func) for name, o in obj.items()}
    elif isinstance(obj, (torch.Tensor, torch.nn.Parameter)):
        b = func(obj)
        print("return tensor: ", b.requires_grad)
        return b

def to_mem_format(mem_format, inputs):
    def inner_to_mem_format(obj):
        old_requires_grad = obj.requires_grad
        _tensor = obj.clone().to(memory_format=mem_format).detach().requires_grad_(old_requires_grad)
        print("inner cloned tensor : ", _tensor.requires_grad)
        return _tensor
    print("input requires_grad: ", inputs[0].requires_grad)
    return traverse_obj(inputs, inner_to_mem_format)


super_run = torch._dynamo.optimize("aot_eager")(to_mem_format)

inputs = (torch.randn(4, 3, 4, 4).requires_grad_(True),)

output = super_run(torch.channels_last, inputs)

print("output requires_grad: ", output[0].requires_grad)
```

Resutls:

```
input requires_grad:  True
inner cloned tensor :  True
return tensor:  False
return tuple tensor:  False
output requires_grad:  False
```

### Versions

Latest master.

cc @ezyang @gchanan @zou3519 @msaroufim @wconstab @bdhirsh @anijain2305 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

**Comment by anijain2305** [comment]
cc @bdhirsh

**Comment by anijain2305** [comment]
Turning this into high priority for now.

**Comment by bdhirsh** [comment]
This actually looks like a variant of this issue: https://github.com/pytorch/pytorch/issues/90552

The issue is that when we call `x.detach.requires_grad_(True)` on a graph intermediate, we're making `x` a "leaf" in the autograd graph, that is not a graph input.

**Comment by bdhirsh** [comment]
As discussed in that issue: a partial fix would be to figure out a way to "lift" graph intermediates into graph inputs, when we see that they require gradients and participate in the output compute. That wouldn't fix all silent correctness issues mentioned (especially for users that do interesting things with `autograd.grad()`, but I think it would fix this particular issue.

**Comment by bdhirsh** [comment]
@eellison pointed out that a simpler fix would be to graph break on the unsupported pattern (in this case, the `requires_grad_(True)` call on the intermediate).

**Comment by eellison** [comment]
See, similar prs here: https://github.com/pytorch/pytorch/pull/108182 https://github.com/pytorch/pytorch/pull/98279
